File: MinStack.cs
    public class MinStack
    {
        // Time Complexity : Push - O(1), Pop - O(1), Top - O(1) and GetMin - O(1)
        // Space Complexity : since we are using 2 stacks- O(N)
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No

        private Stack<int> stack;
        private Stack<int> minStack;
        private int min;

        public MinStack()
        {
            min = int.MaxValue;
            stack = new Stack<int>();
            minStack = new Stack<int>();
            minStack.Push(min);
        }

        public void Push(int val)
        {
            min = Math.Min(min, val);
            stack.Push(val);
            minStack.Push(min);
        }

        public void Pop()
        {
            stack.Pop();
            minStack.Pop();
            min = minStack.Peek();
        }

        public int Top()
        {
            return stack.Peek();
        }

        public int GetMin()
        {
            return minStack.Peek();
        }
    }

File: Min Stack.java
// Time Complexity : push: O(1), pop: O(1), top: O(1), getMin: O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// Maintain a variable min which will have current minimum values
// In case of push, If the incoming element values is lesser than the min value, push the current min and update the min value. Then push the element
// In case of pop, If the popped value is equal to min value, update the min value to top of stack value and pop it.

class MinStack {
    int min;
    Stack<Integer> st;

    public MinStack() {
        min = Integer.MAX_VALUE;
        st = new Stack<>();
    }
    
    public void push(int val) {
        if(val <= min){
            st.push(min);
            min = val;
        }
        st.push(val);
    }
    
    public void pop() {
        if(min == st.pop()){
            min = st.pop();
        }
    }
    
    public int top() {
        return st.peek();
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */